src/components/activity/ActivityDirectiveChangelog.svelte

```diff
@@ -0,0 +1,365 @@
+        ? activityDirectivesMap[current.anchor_id].name
+        : 'Unknown Directive';
+      const previousAnchorDirective = previous.anchor_id
+        ? activityDirectivesMap[previous.anchor_id].name
+        : 'Unknown Directive';
+      differences['Anchored To'] = {
+        currentValue: current.anchor_id ? `${current.anchor_id} - ${currentAnchorDirective}` : 'Plan',
+        previousValue: previous.anchor_id ? `${previous.anchor_id} - ${previousAnchorDirective}` : 'Plan',
+      };
+    }
+
+    if (current.name !== previous.name) {
+      differences['Name'] = {
+        currentValue: current.name,
+        previousValue: previous.name,
+      };
+    }
+
+    const changedProperties = Object.keys(differences);
+    const firstChange = changedProperties.length
+      ? differences[changedProperties[0]]
+      : { currentValue: '', previousValue: '' };
+
+    if (changedProperties.length > 1) {
+      return { currentValue: `${changedProperties.length} Changes`, name: 'Multiple Changes', previousValue: '' };
+    } else if (changedProperties.length === 0) {
+      // Catch edge-case scenarios where there a no differences between revisions
+      // e.g. an API update that didn't actually change anything, or creating a preset from the current parameter values
+      return { currentValue: '', name: 'No Changes', previousValue: '' };
+    }
+
+    return {
+      currentValue: firstChange.currentValue,
+      name: changedProperties[0] || '',
+      previousValue: firstChange.previousValue,
+    };
+  }
+
+  async function restoreRevision(revisionId: number) {
+    const { id: activityId, plan_id: planId } = activityDirective;
+    const restored = await effects.restoreActivityFromChangelog(activityId, planId, revisionId, user);
+
+    if (restored) {
+      dispatch('closeChangelog');
+    }
+  }
+
+  onMount(async () => {
+    const { id: activityId, plan_id: planId } = activityDirective;
+    activityRevisions = await effects.getActivityDirectiveChangelog(planId, activityId, user);
+
+    // Get default set of effective arguments
+    const effectiveDefaultArguments = await effects.getEffectiveActivityArguments(
+      modelId,
+      activityType?.name || '',
+      {},
+      user,
+    );
+
+    // Get effective arguments for all revisions by coalescing defaults with args supplied in revision
+    effectiveRevisionArguments = activityRevisions.map(revision => {
+      return {
+        ...effectiveDefaultArguments?.arguments,
+        ...revision.arguments,
+      };
+    });
+
+    // Build a summary of changes by comparing two revisions and their effective arguments
+    activityRevisionChangeMap = activityRevisions.map((activityRevision, i) => {
+      const previousRevision = activityRevisions[i + 1];
+
+      // At some point there will be no previous revision to compute a difference from
+      if (!previousRevision) {
+        return { currentValue: '', name: 'Earliest Known Revision', previousValue: '' };
```

Comment:
Does `Earliest Known Revision` mean "oldest stored entry in changelog stack" or "original state" of some sort? To me this feels potentially misleading as the user isn't informed that the changelog is restricted to 10 entries?

Reply:
@lklyne

Reply:
Earliest Known Revision is the oldest entry we have available in the changelog.